{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2019 March 7 - HiSeq Genome Skimming Data\n",
    "*****************************\n",
    "\n",
    "## Instruction for data download\n",
    "\n",
    "*****************************\n",
    "\n",
    "### sftp server name: dnaseq2.igsp.duke.edu\n",
    "\n",
    "### Username: zimmer_5477\n",
    "\n",
    "### Password: fhkVwWX3UvLr\n",
    "\n",
    "#### Your data is under zimmer_5477/ZIMMER_5477_190301B1\n",
    "\n",
    "#### Reminder, your data will be available to download until 2019-04-06"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copying files to local HD\n",
    "\n",
    "CWD: /Volumes/Samsung_T5/IsoetesDNA/Illumina/March2019/\n",
    "\n",
    "```\n",
    "sftp zimmer_5477@dnaseq2.igsp.duke.edu\n",
    "zimmer_5477@dnaseq2.igsp.duke.edu's password: \n",
    "Connected to zimmer_5477@dnaseq2.igsp.duke.edu.\n",
    "sftp> cd zimmer_5477/ZIMMER_5477_190301B1\n",
    "sftp> ls -lah\n",
    "drwxr-xr-x    ? 0        501          2.5K Mar  7 04:21 .\n",
    "dr-x------    ? 2762     501           38B Mar  7 04:06 ..\n",
    "-rw-r--r--    ? 0        501          634M Mar  7 04:15 Isoetes_anatolica_S33_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          687M Mar  7 04:09 Isoetes_anatolica_S33_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          2.6G Mar  7 04:12 Isoetes_appalachiana_S_S38_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          2.8G Mar  7 04:20 Isoetes_appalachiana_S_S38_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          2.5G Mar  7 04:11 Isoetes_boomii_S42_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          2.8G Mar  7 04:12 Isoetes_boomii_S42_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.2G Mar  7 04:16 Isoetes_echinospora_1_S30_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.3G Mar  7 04:22 Isoetes_echinospora_1_S30_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.4G Mar  7 04:21 Isoetes_echinospora_2_S31_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.5G Mar  7 04:15 Isoetes_echinospora_2_S31_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          814M Mar  7 04:13 Isoetes_echinospora_3_S32_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          889M Mar  7 04:15 Isoetes_echinospora_3_S32_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.4G Mar  7 04:07 Isoetes_engelmannii_N_S44_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.5G Mar  7 04:19 Isoetes_engelmannii_N_S44_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.2G Mar  7 04:13 Isoetes_georgiana_S41_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.4G Mar  7 04:17 Isoetes_georgiana_S41_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.0G Mar  7 04:11 Isoetes_laurentiana_S45_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.1G Mar  7 04:14 Isoetes_laurentiana_S45_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.7G Mar  7 04:18 Isoetes_microvela_S40_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.8G Mar  7 04:07 Isoetes_microvela_S40_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          754M Mar  7 04:17 Isoetes_pallida_S36_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          826M Mar  7 04:17 Isoetes_pallida_S36_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          550M Mar  7 04:18 Isoetes_prototypus_S27_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          593M Mar  7 04:06 Isoetes_prototypus_S27_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.8G Mar  7 04:11 Isoetes_septentrionalis_S43_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.8G Mar  7 04:16 Isoetes_septentrionalis_S43_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          648M Mar  7 04:19 Isoetes_setacea_S35_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          701M Mar  7 04:21 Isoetes_setacea_S35_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501         1008M Mar  7 04:14 Isoetes_snowii_S28_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.1G Mar  7 04:19 Isoetes_snowii_S28_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.5G Mar  7 04:13 Isoetes_tennesseensis_S37_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.6G Mar  7 04:18 Isoetes_tennesseensis_S37_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          6.2G Mar  7 04:10 Isoetes_texana_S26_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          6.3G Mar  7 04:08 Isoetes_texana_S26_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          499M Mar  7 04:15 Isoetes_toximontana_S34_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          541M Mar  7 04:21 Isoetes_toximontana_S34_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.3G Mar  7 04:16 Isoetes_tuckermanii_S39_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.4G Mar  7 04:18 Isoetes_tuckermanii_S39_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          4.5G Mar  7 04:14 Isoetes_viridimontana_S29_L003_R1_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          4.9G Mar  7 04:21 Isoetes_viridimontana_S29_L003_R2_001.fastq.gz\n",
    "-rw-r--r--    ? 0        501          1.1K Mar  7 04:18 README.rtf\n",
    "-rw-r--r--    ? 0        501          3.9K Mar  7 04:12 ZIMMER_5477_190301B1.checksum\n",
    "```\n",
    "\n",
    "### Download files\n",
    "```\n",
    "get *\n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_anatolica_S33_L003_R1_001.fastq.gz to Isoetes_anatolica_S33_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_anatolica_S33_L003_R1_001.fastq.gz                                                                                                                                                                  100%  634MB   8.0MB/s   01:19    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_anatolica_S33_L003_R2_001.fastq.gz to Isoetes_anatolica_S33_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_anatolica_S33_L003_R2_001.fastq.gz                                                                                                                                                                  100%  687MB   8.1MB/s   01:25    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_appalachiana_S_S38_L003_R1_001.fastq.gz to Isoetes_appalachiana_S_S38_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_appalachiana_S_S38_L003_R1_001.fastq.gz                                                                                                                                                             100% 2706MB   8.0MB/s   05:39    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_appalachiana_S_S38_L003_R2_001.fastq.gz to Isoetes_appalachiana_S_S38_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_appalachiana_S_S38_L003_R2_001.fastq.gz                                                                                                                                                             100% 2906MB   8.0MB/s   06:04    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_boomii_S42_L003_R1_001.fastq.gz to Isoetes_boomii_S42_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_boomii_S42_L003_R1_001.fastq.gz                                                                                                                                                                     100% 2605MB   7.9MB/s   05:30    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_boomii_S42_L003_R2_001.fastq.gz to Isoetes_boomii_S42_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_boomii_S42_L003_R2_001.fastq.gz                                                                                                                                                                     100% 2858MB   8.1MB/s   05:54    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_1_S30_L003_R1_001.fastq.gz to Isoetes_echinospora_1_S30_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_1_S30_L003_R1_001.fastq.gz                                                                                                                                                              100% 1194MB   8.4MB/s   02:22    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_1_S30_L003_R2_001.fastq.gz to Isoetes_echinospora_1_S30_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_1_S30_L003_R2_001.fastq.gz                                                                                                                                                              100% 1292MB   8.4MB/s   02:33    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_2_S31_L003_R1_001.fastq.gz to Isoetes_echinospora_2_S31_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_2_S31_L003_R1_001.fastq.gz                                                                                                                                                              100% 1427MB   8.3MB/s   02:51    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_2_S31_L003_R2_001.fastq.gz to Isoetes_echinospora_2_S31_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_2_S31_L003_R2_001.fastq.gz                                                                                                                                                              100% 1551MB   8.4MB/s   03:04    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_3_S32_L003_R1_001.fastq.gz to Isoetes_echinospora_3_S32_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_3_S32_L003_R1_001.fastq.gz                                                                                                                                                              100%  814MB   8.3MB/s   01:38    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_3_S32_L003_R2_001.fastq.gz to Isoetes_echinospora_3_S32_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_echinospora_3_S32_L003_R2_001.fastq.gz                                                                                                                                                              100%  889MB   8.3MB/s   01:46    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_engelmannii_N_S44_L003_R1_001.fastq.gz to Isoetes_engelmannii_N_S44_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_engelmannii_N_S44_L003_R1_001.fastq.gz                                                                                                                                                              100% 1434MB   8.3MB/s   02:52    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_engelmannii_N_S44_L003_R2_001.fastq.gz to Isoetes_engelmannii_N_S44_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_engelmannii_N_S44_L003_R2_001.fastq.gz                                                                                                                                                              100% 1545MB   8.3MB/s   03:05    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_georgiana_S41_L003_R1_001.fastq.gz to Isoetes_georgiana_S41_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_georgiana_S41_L003_R1_001.fastq.gz                                                                                                                                                                  100% 1271MB   8.3MB/s   02:32    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_georgiana_S41_L003_R2_001.fastq.gz to Isoetes_georgiana_S41_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_georgiana_S41_L003_R2_001.fastq.gz                                                                                                                                                                  100% 1392MB   8.3MB/s   02:47    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_laurentiana_S45_L003_R1_001.fastq.gz to Isoetes_laurentiana_S45_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_laurentiana_S45_L003_R1_001.fastq.gz                                                                                                                                                                100% 1031MB   8.3MB/s   02:04    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_laurentiana_S45_L003_R2_001.fastq.gz to Isoetes_laurentiana_S45_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_laurentiana_S45_L003_R2_001.fastq.gz                                                                                                                                                                100% 1122MB   8.3MB/s   02:14    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_microvela_S40_L003_R1_001.fastq.gz to Isoetes_microvela_S40_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_microvela_S40_L003_R1_001.fastq.gz                                                                                                                                                                  100% 1690MB   8.3MB/s   03:23    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_microvela_S40_L003_R2_001.fastq.gz to Isoetes_microvela_S40_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_microvela_S40_L003_R2_001.fastq.gz                                                                                                                                                                  100% 1849MB   8.3MB/s   03:42    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_pallida_S36_L003_R1_001.fastq.gz to Isoetes_pallida_S36_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_pallida_S36_L003_R1_001.fastq.gz                                                                                                                                                                    100%  754MB   8.3MB/s   01:31    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_pallida_S36_L003_R2_001.fastq.gz to Isoetes_pallida_S36_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_pallida_S36_L003_R2_001.fastq.gz                                                                                                                                                                    100%  826MB   8.3MB/s   01:39    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_prototypus_S27_L003_R1_001.fastq.gz to Isoetes_prototypus_S27_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_prototypus_S27_L003_R1_001.fastq.gz                                                                                                                                                                 100%  550MB   8.3MB/s   01:06    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_prototypus_S27_L003_R2_001.fastq.gz to Isoetes_prototypus_S27_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_prototypus_S27_L003_R2_001.fastq.gz                                                                                                                                                                 100%  592MB   8.2MB/s   01:12    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_septentrionalis_S43_L003_R1_001.fastq.gz to Isoetes_septentrionalis_S43_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_septentrionalis_S43_L003_R1_001.fastq.gz                                                                                                                                                            100% 1837MB   7.8MB/s   03:55    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_septentrionalis_S43_L003_R2_001.fastq.gz to Isoetes_septentrionalis_S43_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_septentrionalis_S43_L003_R2_001.fastq.gz                                                                                                                                                            100% 1877MB   7.9MB/s   03:57    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_setacea_S35_L003_R1_001.fastq.gz to Isoetes_setacea_S35_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_setacea_S35_L003_R1_001.fastq.gz                                                                                                                                                                    100%  648MB   7.9MB/s   01:21    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_setacea_S35_L003_R2_001.fastq.gz to Isoetes_setacea_S35_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_setacea_S35_L003_R2_001.fastq.gz                                                                                                                                                                    100%  700MB   7.9MB/s   01:28    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_snowii_S28_L003_R1_001.fastq.gz to Isoetes_snowii_S28_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_snowii_S28_L003_R1_001.fastq.gz                                                                                                                                                                     100% 1008MB   8.0MB/s   02:06    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_snowii_S28_L003_R2_001.fastq.gz to Isoetes_snowii_S28_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_snowii_S28_L003_R2_001.fastq.gz                                                                                                                                                                     100% 1110MB   8.0MB/s   02:19    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tennesseensis_S37_L003_R1_001.fastq.gz to Isoetes_tennesseensis_S37_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tennesseensis_S37_L003_R1_001.fastq.gz                                                                                                                                                              100% 1550MB   8.0MB/s   03:14    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tennesseensis_S37_L003_R2_001.fastq.gz to Isoetes_tennesseensis_S37_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tennesseensis_S37_L003_R2_001.fastq.gz                                                                                                                                                              100% 1681MB   8.0MB/s   03:29    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_texana_S26_L003_R1_001.fastq.gz to Isoetes_texana_S26_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_texana_S26_L003_R1_001.fastq.gz                                                                                                                                                                     100% 6310MB   7.9MB/s   13:14    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_texana_S26_L003_R2_001.fastq.gz to Isoetes_texana_S26_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_texana_S26_L003_R2_001.fastq.gz                                                                                                                                                                     100% 6432MB   8.1MB/s   13:18    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_toximontana_S34_L003_R1_001.fastq.gz to Isoetes_toximontana_S34_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_toximontana_S34_L003_R1_001.fastq.gz                                                                                                                                                                100%  499MB   7.9MB/s   01:03    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_toximontana_S34_L003_R2_001.fastq.gz to Isoetes_toximontana_S34_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_toximontana_S34_L003_R2_001.fastq.gz                                                                                                                                                                100%  541MB   8.0MB/s   01:07    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tuckermanii_S39_L003_R1_001.fastq.gz to Isoetes_tuckermanii_S39_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tuckermanii_S39_L003_R1_001.fastq.gz                                                                                                                                                                100% 1301MB   8.0MB/s   02:43    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tuckermanii_S39_L003_R2_001.fastq.gz to Isoetes_tuckermanii_S39_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_tuckermanii_S39_L003_R2_001.fastq.gz                                                                                                                                                                100% 1418MB   8.0MB/s   02:57    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_viridimontana_S29_L003_R1_001.fastq.gz to Isoetes_viridimontana_S29_L003_R1_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_viridimontana_S29_L003_R1_001.fastq.gz                                                                                                                                                              100% 4596MB   8.0MB/s   09:34    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/Isoetes_viridimontana_S29_L003_R2_001.fastq.gz to Isoetes_viridimontana_S29_L003_R2_001.fastq.gz\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/Isoetes_viridimontana_S29_L003_R2_001.fastq.gz                                                                                                                                                              100% 5001MB   7.0MB/s   11:55    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/README.rtf to README.rtf\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/README.rtf                                                                                                                                                                                                  100% 1126    15.0KB/s   00:00    \n",
    "Fetching /zimmer_5477/ZIMMER_5477_190301B1/ZIMMER_5477_190301B1.checksum to ZIMMER_5477_190301B1.checksum\n",
    "/zimmer_5477/ZIMMER_5477_190301B1/ZIMMER_5477_190301B1.checksum                                                                                                                                                                               100% 3978    51.5KB/s   00:00    \n",
    "```\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copying files to ODU Turing Cluster\n",
    "\n",
    "CWD: /scratch-lustre/pscha005/Illumina/\n",
    "\n",
    "Copied to cluster from Duke server as above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing script to automate plastome assembly: Trimmomatic -> Bowtie2 -> SPAdes\n",
    "\n",
    "### makePlastome.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**********\n",
      "WARNING: No reference genome provided. SPAdes will be run on all data.\n",
      "**********\n",
      "!!!!!!!!!!\n",
      "ERROR: No data (fastq files) supplied\n",
      "!!!!!!!!!!\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Peter/.local/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2889: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#! /usr/bin/python\n",
    "\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "sampleDict = {} # Structure is {'Sample1':['Input_File_R1.fastq.gz','Input_File_R2.fastq.gz'], 'Sample2'...}\n",
    "hpc = 0\n",
    "fileList= []\n",
    "if \"-hpc\" in sys.argv:\n",
    "\thpc = 1\n",
    "for item in sys.argv:\n",
    "\tif \".fastq\" in item: \n",
    "\t\tfileList.append(item)\n",
    "\tif \".fasta\" in item:\n",
    "\t\treferenceFile = item\n",
    "\t\treferenceName = \".\".join(referenceFile.split(\".\")[:-1])\n",
    "\n",
    "try:\n",
    "\tif len(referenceFile) == 1:\n",
    "\t\tprint \"Using reference: %s\" %(referenceFile)\n",
    "except:\n",
    "\tprint \"*\" *10\n",
    "\tprint \"WARNING: No reference genome provided. SPAdes will be run on all data.\"\n",
    "\tprint \"*\" *10\n",
    "\n",
    "if len(fileList) == 0:\n",
    "\tprint \"!\" *10\n",
    "\tprint \"ERROR: No data (fastq files) supplied\"\n",
    "\tprint \"!\" *10\n",
    "\tsys.exit()\n",
    "\n",
    "for fileName in fileList:\n",
    "\tsplitFileName = fileName.strip(\"\\n\").split(\"_\")\n",
    "\tsampleName = \"_\".join(splitFileName[:-2])\n",
    "\tsampleDict[sampleName] = []\n",
    "\t\n",
    "for fileName in fileList:\n",
    "\tsplitFileName = fileName.strip(\"\\n\").split(\"_\")\n",
    "\tsampleName = \"_\".join(splitFileName[:-2])\n",
    "\tsampleDict[sampleName].append(fileName)\n",
    "\n",
    "if hpc == 1:\n",
    "\tprint \"*\"*10\n",
    "\tprint \"Running in HPC mode...\"\n",
    "\n",
    "'''Trimmomatic'''\n",
    "def trimmomatic():\n",
    "\tprint \"*\"*10\n",
    "\tprint \"Running trimmomatic...\"\n",
    "\tprint \"*\"*10\n",
    "\tglobal trimOutputDict\n",
    "\ttrimOutputDict = {}\n",
    "\tfor sample in sorted(sampleDict.keys()):\n",
    "\t\tsubprocess.call(['mkdir', sample])\n",
    "\t\tif len(sampleDict[sample]) == 1 and hpc == 0:\n",
    "\t\t\tprint \"*\" * 10\n",
    "\t\t\tprint \"WARNING: %s does not have 2 input files. Trimmomatic will be run in SE mode\" %(sample)\n",
    "\t\t\tprint \"*\" * 10\n",
    "\t\t\ttry:\n",
    "\t\t\t\tsubprocess.call(['java', '-jar', 'trimmomatic-0.33.jar', 'SE', '-phred33', '%s' %(sampleDict[sample][0]), './%s/%s_TRIM.fq.gz'%(sample,sample), 'ILLUMINACLIP:./adapters/TruSeq3-SE.fa:2:30:10', 'LEADING:3', 'TRAILING:3', 'SLIDINGWINDOW:4:15', 'MINLEN:36'])\n",
    "\t\t\texcept:\n",
    "\t\t\t\tsys.exit(\"ERROR: Trimmomatic.jar not found in current directory\")\n",
    "\t\t\ttrimOutputDict[sample] = \"%s_TRIM.fq.gz\" %(sample)\n",
    "\t\telif len(sampleDict[sample]) == 1 and hpc == 1:\n",
    "\t\t\ttrimOutputDict[sample] = \"%s_TRIM.fq.gz\" %(sample)\n",
    "\t\t\ttrimScript = open(\"trimmomatic_%s.sh\" %(sample), \"w\")\n",
    "\t\t\ttrimScript.write(\n",
    "'''#!/bin/bash -l\\n\n",
    "#SBATCH --job-name=trim    # Job name\n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --output=trimmomatic_%s.out   # Standard output and error log\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "enable_lmod\n",
    "module load java/11.0\n",
    "java -jar /scratch-lustre/pscha005/trimmomatic/trimmomatic-0.33.jar SE -threads 16 %s ./%s/%s_TRIM.fq.gz ILLUMINACLIP:/scratch-lustre/pscha005/trimmomatic/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
    "date\n",
    "'''%(sample, sampleDict[sample][0], sample, sample))\n",
    "\t\t\ttrimScript.close()\n",
    "\t\telif len(sampleDict[sample]) == 2 and hpc == 0:\n",
    "\t\t\ttry:\n",
    "\t\t\t\tsubprocess.call(['java', '-jar', 'trimmomatic-0.33.jar', 'PE', '-phred33', '%s' %(sampleDict[sample][0]), '%s' %(sampleDict[sample][1]), './%s/%s_TRIM_forward_paired.fq.gz'%(sample,sample), './%s/%s_TRIM_forward_unpaired.fq.gz'%(sample,sample), './%s/%s_TRIM_reverse_paired.fq.gz'%(sample,sample), './%s/%s_TRIM_reverse_unpaired.fq.gz'%(sample,sample), 'ILLUMINACLIP:./adapters/TruSeq3-PE.fa:2:30:10', 'LEADING:3', 'TRAILING:3', 'SLIDINGWINDOW:4:15', 'MINLEN:36'])\n",
    "\t\t\texcept:\n",
    "\t\t\t\tsys.exit(\"ERROR: Trimmomatic.jar not found in current directory\")\n",
    "\t\t\ttrimOutputDict[sample] = ['%s_TRIM_forward_paired.fq.gz'%(sample), '%s_TRIM_reverse_paired.fq.gz'%(sample)]\n",
    "\t\telif len(sampleDict[sample]) == 2 and hpc == 1:\n",
    "\t\t\ttrimOutputDict[sample] = ['%s_TRIM_forward_paired.fq.gz'%(sample), '%s_TRIM_reverse_paired.fq.gz'%(sample)]\n",
    "\t\t\ttrimScript = open(\"trimmomatic_%s.sh\" %(sample), \"w\")\n",
    "\t\t\ttrimScript.write(\n",
    "'''#!/bin/bash -l\\n\n",
    "#SBATCH --job-name=trim    # Job name\n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --output=trimmomatic_%s.out   # Standard output and error log\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "enable_lmod\n",
    "module load java/11.0\n",
    "java -jar /scratch-lustre/pscha005/trimmomatic/trimmomatic-0.33.jar PE -threads 16 %s %s ./%s/%s_TRIM_forward_paired.fq.gz ./%s/%s_TRIM_forward_unpaired.fq.gz ./%s/%s_TRIM_reverse_paired.fq.gz ./%s/%s_TRIM_reverse_unpaired.fq.gz ILLUMINACLIP:/scratch-lustre/pscha005/trimmomatic/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
    "date\n",
    "'''%(sample, sampleDict[sample][0], sampleDict[sample][1], sample, sample, sample, sample, sample, sample, sample, sample))\n",
    "\t\t\ttrimScript.close()\n",
    "\n",
    "\t\telif len(sampleDict[sample]) > 2:\n",
    "\t\t\tprint \"!\" * 10\n",
    "\t\t\tprint \"ERROR: Too many files specified for %s\" %(sample)\n",
    "\t\t\tprint \"!\" * 10\n",
    "\t\t\tsys.exit()\n",
    "\n",
    "'''Bowtie2'''\n",
    "def bowtie():\n",
    "\tif 'referenceName' in globals():\n",
    "\t\tprint \"*\"*10\n",
    "\t\tprint \"Running bowtie...\"\n",
    "\t\tprint \"*\"*10\n",
    "\n",
    "\t\treference = \"reference\"\n",
    "\t\tsubprocess.call(['mkdir', reference])\n",
    "\t\treferenceOutput = \"./reference/%s\" %(referenceName)\n",
    "\t\ttry:\n",
    "\t\t\tsubprocess.call(['bowtie2-build', '-q', referenceFile, referenceOutput])\n",
    "\t\texcept:\n",
    "\t\t\tsys.exit(\"ERROR: bowtie2 not found in PATH\")\n",
    "\t\tglobal bowtieOutputDict\n",
    "\t\tbowtieOutputDict = {}\n",
    "\t\tfor sample in sorted(trimOutputDict.keys()):\n",
    "\t\t\tif len(trimOutputDict[sample]) == 1 and hpc == 0:\n",
    "\t\t\t\tsubprocess.call(['bowtie2','--quiet', '--very-sensitive-local', '-x', referenceOutput, '-U', './%s/%s' %(sample,trimOutputDict[sample][0]), '--al-gz', './%s/%s_align.fq.gz' %(sample,sample), '--un-gz', './%s/%s_unalign.fq.gz' %(sample,sample)])\n",
    "\t\t\t\tbowtieOutputDict[sample] = [\"%s_align.fq.gz\" %(sample), \"%s_unalign.fq.gz\" %(sample)]\n",
    "\t\t\telif len(trimOutputDict[sample]) == 1 and hpc == 1:\n",
    "\t\t\t\tbowtieOutputDict[sample] = [\"%s_align.fq.gz\" %(sample), \"%s_unalign.fq.gz\" %(sample)]\n",
    "\t\t\t\tbowtieScript = open(\"bowtie_%s.sh\" %(sample), \"w\")\n",
    "\t\t\t\tbowtieScript.write(\n",
    "'''#!/bin/bash\\n\n",
    "#SBATCH --job-name=bwtie    # Job name\n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "module load bowtie2/2.2.4\n",
    "bowtie2 --quiet --very-sensitive-local -x %s -U ./%s/%s --al-gz ./%s/%s_align.fq.gz --un-gz ./%s/%s_unalign.fq.gz\n",
    "\n",
    "date\n",
    "'''%(referenceOutput, sample, trimOutputDict[sample][0], sample, sample, sample, sample))\n",
    "\t\t\t\tbowtieScript.close()\n",
    "\t\t\telif len(trimOutputDict[sample]) == 2 and hpc == 0:\n",
    "\t\t\t\tsubprocess.call(['bowtie2','--quiet', '--very-sensitive-local', '-x', referenceOutput, '-1', './%s/%s' %(sample,trimOutputDict[sample][0]), '-2', './%s/%s' %(sample,trimOutputDict[sample][1]), '--al-conc-gz', './%s/%s_R%%_align.fq.gz' %(sample,sample), '--un-conc-gz', './%s/%s_R%%_unalign.fq.gz' %(sample,sample)])\n",
    "\t\t\t\tbowtieOutputDict[sample] = [\"%s_R1_align.fq\" %(sample), \"%s_R2_align.fq\" %(sample), \"%s_R1_unalign.fq\" %(sample), \"%s_R2_unalign.fq\" %(sample)]\n",
    "\t\t\telif len(trimOutputDict[sample]) == 2 and hpc == 1:\n",
    "\t\t\t\tbowtieOutputDict[sample] = [\"%s_R1_align.fq\" %(sample), \"%s_R2_align.fq\" %(sample), \"%s_R1_unalign.fq\" %(sample), \"%s_R2_unalign.fq\" %(sample)]\n",
    "\t\t\t\tbowtieScript = open(\"bowtie_%s.sh\" %(sample), \"w\")\n",
    "\t\t\t\tbowtieScript.write(\n",
    "'''#!/bin/bash\\n\n",
    "#SBATCH --job-name=bwtie    # Job name\n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "module load bowtie2/2.2.4\n",
    "bowtie2 --quiet --very-sensitive-local -x %s -1 ./%s/%s -2 ./%s/%s --al-conc-gz ./%s/%s_R%%_align.fq.gz --un-conc-gz ./%s/%s_R%%_unalign.fq.gz\n",
    "\n",
    "date\n",
    "'''%(referenceOutput, sample, trimOutputDict[sample][0], sample, trimOutputDict[sample][1], sample, sample, sample, sample))\n",
    "\t\t\tbowtieScript.close()\n",
    "\telse:\n",
    "\t\tprint \"*\"*10\n",
    "\t\tprint \"WARNING: No reference file found, skipping bowtie filtering...\"\n",
    "\t\tprint \"*\"*10\n",
    "\n",
    "\n",
    "'''SPAdes'''\n",
    "def spades():\n",
    "\tfor sample in sorted(bowtieOutputDict.keys()):\n",
    "\t\tif hpc == 0 and len(bowtieOutputDict[sample]) == 2:\n",
    "\t\t\tsubprocess.call(['spades.py', '-k', '21,33,55,77', '--careful', '-s', './%s/%s' %(sample, bowtieOutputDict[sample][0]),'-o', './%s/spades_%s' %(sample, sample)])\n",
    "\t\telif hpc == 1 and len(bowtieOutputDict[sample]) == 2:\n",
    "\t\t\tspadesScript = open(\"spades_%s.sh\" %(sample), \"w\")\n",
    "\t\t\tspadesScript.write(\n",
    "'''#!/bin/bash -l\n",
    "#SBATCH --job-name=spds    # Job name\n",
    "#SBATCH --mail-type=ALL               # Mail events (NONE, BEGIN, END, FAIL, ALL)\n",
    "#SBATCH --mail-user=<pscha005@odu.edu>   # Where to send mail   \n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --output=spades_%s.out   # Standard output and error log\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "enable_lmod\n",
    "module load spades/3.13\n",
    "spades.py -t 16 -k 21,33,55,77 --careful -s ./%s/%s -o ./%s/spades_%s\n",
    "\n",
    "date\n",
    "'''%(sample, sample, bowtieOutputDict[sample][0], sample, bowtieOutputDict[sample][1], sample, sample))\n",
    "\t\t\tspadesScript.close()\n",
    "\t\telif hpc == 0 and len(bowtieOutputDict[sample]) == 4:\n",
    "\t\t\tsubprocess.call(['spades.py', '-k', '21,33,55,77', '--careful', '-1', './%s/%s' %(sample, bowtieOutputDict[sample][0]), '-2', './%s/%s' %(sample, bowtieOutputDict[sample][1]),'-o', './%s/spades_%s' %(sample, sample)])\n",
    "\t\telif hpc == 1 and len(bowtieOutputDict[sample]) == 4:\n",
    "\t\t\tspadesScript = open(\"spades_%s.sh\" %(sample), \"w\")\n",
    "\t\t\tspadesScript.write(\n",
    "'''#!/bin/bash -l\n",
    "#SBATCH --job-name=spds    # Job name\n",
    "#SBATCH --mail-type=ALL               # Mail events (NONE, BEGIN, END, FAIL, ALL)\n",
    "#SBATCH --mail-user=<pscha005@odu.edu>   # Where to send mail   \n",
    "#SBATCH --ntasks=1                    # Run on a single CPU\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --output=spades_%s.out   # Standard output and error log\n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "enable_lmod\n",
    "module load spades/3.13\n",
    "spades.py -t 16 -k 21,33,55,77 --careful -1 ./%s/%s -2 ./%s/%s -o ./%s/spades_%s\n",
    "\n",
    "date\n",
    "'''%(sample, sample, bowtieOutputDict[sample][0], sample, bowtieOutputDict[sample][1], sample, sample))\n",
    "\t\t\tspadesScript.close()\n",
    "\t\t\t\n",
    "trimmomatic()\n",
    "bowtie()\n",
    "spades()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ran makePlastome script to generate SLURM scripts for all samples\n",
    "\n",
    "### 1st step: Trimmomatic scripts submitted on Turing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pwd\n",
    "/scratch-lustre/pscha005/Illumina\n",
    "bash\n",
    "for i in trimmomatic*.sh; do sbatch $i; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trimmomatic results analyzed using FastQC on Turing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "module load fastqc/0.11.5\n",
    "module load java/1.8\n",
    "for i in ./*/*paired.fq.gz; do fastqc $i; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2nd step: Bowtie2 scripts submitted on Turing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pwd\n",
    "/scratch-lustre/pscha005/Illumina\n",
    "bash\n",
    "for i in spades*.sh; do sbatch $i; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
